### ISSUE TITLE
Strict parsing for IntEnum

### ISSUE THREAD

**Issue by sbrandtb** [pydantic/pydantic#1841]
# Feature Request

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.5.1
            pydantic compiled: True
                 install path: <snip>
               python version: 3.7.5 (default, Oct 15 2019, 22:40:07)  [GCC 7.4.0]
                     platform: Linux-4.15.0-112-generic-x86_64-with-Ubuntu-18.04-bionic
     optional deps. installed: ['typing-extensions']

```

It looks like currently Pydantic does not allow strict parsing of `IntEnum`s. The non-strict int validator is hardcoded to `validators.py` `_VALIDATORS`.

I would like to be able to enable strict parsing, i.e.:

```
from enum import IntEnum
import pydantic


class SomeIntEnum(IntEnum):
    foo = 1
    bar = 2


class SomeModel(BaseModel):
    thing: SomeIntEnum
```

I could imagine a parameter to `Field`, with a superclass `StrictIntEnum` of Pydantic (which would possibly be the simplest solution) or a magic field in the `IntEnum` class, like `__strict__`. Happy to make a PR after agreeing on which way to go.

**Comment by PrettyWood** [comment]
Hello @sbrandtb 
In the meantime if you want to have a strict enum, you can just do

```python
from enum import Enum

from pydantic import BaseModel


class StrictEnum(Enum):
    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, value):
        if not isinstance(value, cls):
            raise TypeError(f'{value!r} should be an {cls.__name__} value')
        return value


class FooEnum(int, StrictEnum):
    foo = 1
    bar = 2

class SomeModel(BaseModel):
    thing: FooEnum

SomeModel(thing=FooEnum.foo)
# thing=<FooEnum.foo: 1>
SomeModel(thing=1)
# thing
#   1 should be an FooEnum value (type=type_error)
```

Hope it helps

**Comment by sbrandtb** [comment]
@PrettyWood Yeah, that looks promising. I'll try it out tomorrow. I was not aware that it was so easy to create proper type classes myself :)

**Comment by PrettyWood** [comment]
I reminds me of https://github.com/samuelcolvin/pydantic/issues/1763#issuecomment-664240687.
I feel like this should be added in *pydantic* one way or another.

Here you would just write
```python
class SomeModel(BaseModel):
    thing: instance_of(SomeIntEnum)
```
and you would have the desired behaviour.

**Comment by sbrandtb** [comment]
For the record: Since I wanted to make sure that for an `IntEnum`, only `int`s are accepted in the passed json/dict, but the above solution validates that those values would be instances of the enum, my solution is following:

```py
class StrictIntEnum(IntEnum):
    @classmethod
    def __get_validators__(cls) -> Generator[AnyCallable, None, None]:
        yield strict_int_validator
        yield enum_validator
```

**Comment by dmontagu** [comment]
V2 now lets you do strict validation of this, in a way that works with json:

```python
from enum import IntEnum

from pydantic import BaseModel, ValidationError


class MyEnum(IntEnum):
    x = 1
    y = 2


class DemoModel(BaseModel):
    e: MyEnum

    model_config = {"strict": True}


try:
    DemoModel(e='1')
except ValidationError as e:
    print(e)
#> Input should be an instance of MyEnum [type=is_instance_of, input_value=1, input_type=int]


model = DemoModel.model_validate_json('{"e":1}')  # no error

try:
    DemoModel.model_validate_json('{"e":"1"}')
except ValidationError as e:
    print(e)
#> Input should be an instance of MyEnum [type=is_instance_of, input_value='1', input_type=str]
```

I think we could improve the error a bit when parsing enum from json, but I will close this issue, and one can be created for that if desired.

**Comment by raulparada** [comment]
Hi. Not sure why this was closed as still relevant; the example provided:
```py
model = DemoModel.model_validate_json('{"e":1}')  # no error
```
indeed works, but I'd also expect the equivalent(?):
```py
model = DemoModel.model_validate({"e":1}) # so, the dict from that same json
```
to give the same result, but this one throws the `should be an instance of MyEnum` validation error. That seems pretty inconsistent to me?

**Comment by bhelabhav** [comment]
> V2 now lets you do strict validation of this, in a way that works with json:
> 
> ```python
> from enum import IntEnum
> 
> from pydantic import BaseModel, ValidationError
> 
> 
> class MyEnum(IntEnum):
>     x = 1
>     y = 2
> 
> 
> class DemoModel(BaseModel):
>     e: MyEnum
> 
>     model_config = {"strict": True}
> 
> 
> try:
>     DemoModel(e='1')
> except ValidationError as e:
>     print(e)
> #> Input should be an instance of MyEnum [type=is_instance_of, input_value=1, input_type=int]
> 
> 
> model = DemoModel.model_validate_json('{"e":1}')  # no error
> 
> try:
>     DemoModel.model_validate_json('{"e":"1"}')
> except ValidationError as e:
>     print(e)
> #> Input should be an instance of MyEnum [type=is_instance_of, input_value='1', input_type=str]
> ```
> 
> I think we could improve the error a bit when parsing enum from json, but I will close this issue, and one can be created for that if desired.

Still seems to error on my end with this example:
```
$ uv run test-pydantic-strict-2.py
1 validation error for DemoModel
e
  Input should be an instance of MyEnum [type=is_instance_of, input_value='1', input_type=str]
    For further information visit https://errors.pydantic.dev/2.10/v/is_instance_of
1 validation error for DemoModel
e
  Input should be 1 or 2 [type=enum, input_value='1', input_type=str]
    For further information visit https://errors.pydantic.dev/2.10/v/enum




$ cat test-pydantic-strict-2.py
from enum import IntEnum

from pydantic import BaseModel, ValidationError


class MyEnum(IntEnum):
    x = 1
    y = 2


class DemoModel(BaseModel):
    e: MyEnum

    model_config = {"strict": True}


try:
    DemoModel(e='1')
except ValidationError as e:
    print(e)
#> Input should be an instance of MyEnum [type=is_instance_of, input_value=1, input_type=int]


model = DemoModel.model_validate_json('{"e":1}')  # no error

try:
    DemoModel.model_validate_json('{"e":"1"}')
except ValidationError as e:
    print(e)
#> Input should be an instance of MyEnum [type=is_instance_of, input_value='1', input_type=str]




$ $ uv tree
Resolved 5 packages in 0.31ms
pydantic-field-serializer v0.1.0
└── pydantic v2.10.4
    ├── annotated-types v0.7.0
    ├── pydantic-core v2.27.2
    │   └── typing-extensions v4.12.2
    └── typing-extensions v4.12.2
```
